File: backend/app/services/bloom_filter.py
"""
Bloom Filter utilities
"""
import math


class BloomFilter:
    """Bloom filter implementation"""
    def __init__(self, size: int, hash_count: int):
        self._size = size
        self._hash_count = hash_count
        self._bits = [False] * size

    def add(self, item: str) -> None:
        """Add an item"""
        for i in range(self._hash_count):
            index = self._hash(item, i) % self._size
            self._bits[index] = True

    def contains(self, item: str) -> bool:
        """Check if item might exist"""
        for i in range(self._hash_count):
            index = self._hash(item, i) % self._size
            if not self._bits[index]:
                return False
        return True

    def _hash(self, item: str, seed: int) -> int:
        """Hash function"""
        hash_val = 0
        for char in item:
            hash_val = ((hash_val << 5) - hash_val + ord(char) + seed) & 0xFFFFFFFF
        return abs(hash_val)

    def clear(self) -> None:
        """Clear all bits"""
        self._bits = [False] * self._size

    def get_false_positive_rate(self) -> float:
        """Calculate false positive rate"""
        set_bits = sum(self._bits)
        return (set_bits / self._size) ** self._hash_count

    @staticmethod
    def optimal_size(expected_items: int, fp_rate: float) -> tuple[int, int]:
        """Calculate optimal size and hash count"""
        size = int(-expected_items * math.log(fp_rate) / (math.log(2) ** 2))
        hash_count = int((size / expected_items) * math.log(2))
        return size, max(1, hash_count)

File: src/lib/bloom-filter.ts
/**
 * Bloom Filter utilities
 */

export class BloomFilter {
  private bits: boolean[]
  private size: number
  private hashCount: number

  constructor(size: number, hashCount: number) {
    this.size = size
    this.hashCount = hashCount
    this.bits = new Array(size).fill(false)
  }

  add(item: string): void {
    for (let i = 0; i < this.hashCount; i++) {
      const index = this.hash(item, i) % this.size
      this.bits[index] = true
    }
  }

  contains(item: string): boolean {
    for (let i = 0; i < this.hashCount; i++) {
      const index = this.hash(item, i) % this.size
      if (!this.bits[index]) {
        return false
      }
    }
    return true
  }

  private hash(item: string, seed: number): number {
    let hash = 0
    for (let i = 0; i < item.length; i++) {
      const char = item.charCodeAt(i)
      hash = ((hash << 5) - hash + char + seed) | 0
    }
    return Math.abs(hash)
  }

  clear(): void {
    this.bits.fill(false)
  }

  getFalsePositiveRate(): number {
    const setBits = this.bits.filter(b => b).length
    return Math.pow(setBits / this.size, this.hashCount)
  }
}
